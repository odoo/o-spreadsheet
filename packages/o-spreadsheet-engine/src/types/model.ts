import { GeoChartRegion } from "./chart/geo_chart";
import { Client, ClientPosition } from "./collaborative/session";
import { TransportService } from "./collaborative/transport_service";
import { Currency } from "./currency";
import { InformationNotification } from "./env";
import { FileStore } from "./files";
import { Locale } from "./locale";
import { Color } from "./misc";

/**
 * Model
 *
 * The Model class is the owner of the state of the Spreadsheet. However, it
 * has more a coordination role: it defers the actual state manipulation work to
 * plugins.
 *
 * At creation, the Model instantiates all necessary plugins. They each have
 * a private state (for example, the Selection plugin has the current selection).
 *
 * State changes are then performed through commands.  Commands are dispatched
 * to the model, which will then relay them to each plugins (and the history
 * handler). Then, the model will trigger an 'update' event to notify whoever
 * is concerned that the command was applied (if it was not cancelled).
 *
 * Also, the model has an unconventional responsibility: it actually renders the
 * visible viewport on a canvas. This is because each plugins actually manage a
 * specific concern about the content of the spreadsheet, and it is more natural
 * if they are able to read data from their internal state to represent it on the
 * screen.
 *
 * Note that the Model can be used in a standalone way to manipulate
 * programmatically a spreadsheet.
 */
export type Mode = "normal" | "readonly" | "dashboard";

export interface ModelConfig {
  readonly mode: Mode;
  /**
   * Any external custom dependencies your custom plugins or functions might need.
   * They are available in plugins config and functions
   * evaluation context.
   */
  readonly custom: Readonly<{
    [key: string]: any;
  }>;
  readonly defaultCurrency?: Partial<Currency>;
  /**
   * External dependencies required to enable some features
   * such as uploading images.
   */
  readonly external: Readonly<ModelExternalConfig>;
  readonly moveClient: (position: ClientPosition) => void;
  readonly transportService: TransportService;
  readonly client: Client;
  readonly snapshotRequested: boolean;
  readonly notifyUI: (payload: InformationNotification) => void;
  readonly raiseBlockingErrorUI: (text: string) => void;
  readonly customColors: Color[];
}

export interface ModelExternalConfig {
  readonly fileStore?: FileStore;
  readonly loadCurrencies?: () => Promise<Currency[]>;
  readonly loadLocales?: () => Promise<Locale[]>;
  readonly geoJsonService?: {
    getAvailableRegions: () => GeoChartRegion[];
    getTopoJson: (region: string) => Promise<any>;
    /**  Convert the name of a geographical feature (eg. France) to the id of the corresponding feature in the TopoJSON */
    geoFeatureNameToId: (region: string, territory: string) => string | undefined;
  };
}
