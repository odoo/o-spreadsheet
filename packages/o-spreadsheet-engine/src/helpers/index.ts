export * from "./misc";
export * from "./numbers";
export * from "./references";
export * from "./uuid";
