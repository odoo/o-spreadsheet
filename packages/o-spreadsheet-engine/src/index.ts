export { FunctionCodeBuilder, Scope, type FunctionCode } from "./formulas/code_builder";
export {
  OPERATOR_MAP,
  UNARY_OPERATOR_MAP,
  compile,
  compileTokens,
  functionCache,
} from "./formulas/compiler";
export {
  OP_PRIORITY,
  convertAstNodes,
  iterateAstNodes,
  mapAst,
  parse,
  parseTokens,
} from "./formulas/parser";
export type {
  AST,
  ASTFuncall,
  ASTOperation,
  ASTString,
  ASTSymbol,
  ASTUnaryOperation,
} from "./formulas/parser";
export { rangeTokenize } from "./formulas/range_tokenizer";
export { POSTFIX_UNARY_OPERATORS, tokenize } from "./formulas/tokenizer";
export type { Token } from "./formulas/tokenizer";
export { FunctionRegistry } from "./functions/function_registry";
export { categories } from "./functions/functionRegistry";
export * from "./helpers";
export { Model } from "./model";
export { BasePlugin } from "./plugins/base_plugin";
export { Registry } from "./registry";
export { StateObserver, type StateObserverChange } from "./state_observer";
export * from "./translation";
export { CellValue } from "./types/cells";
export { SpreadsheetClipboardData } from "./types/clipboard";
export { CoreGetters, PluginGetters } from "./types/coreGetters";
export * from "./types/errors";
export { Format } from "./types/format";
export * from "./types/history";
export * from "./types/locale";
export * from "./types/misc";
export * from "./types/validator";
export const __info__ = {};
