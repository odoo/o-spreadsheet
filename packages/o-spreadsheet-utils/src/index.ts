export * from "./translation";
export * from "./uuid";
