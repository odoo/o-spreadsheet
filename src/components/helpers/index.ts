export * from "@odoo/o-spreadsheet-engine/components/helpers/css";
