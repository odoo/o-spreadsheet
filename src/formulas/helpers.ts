import { functionRegistry } from "../functions";
import { AST, ASTFuncall, iterateAstNodes, parseTokens } from "./parser";
import { Token } from "./tokenizer";

const functions = functionRegistry.content;

export function isExportableToExcel(tokens: Token[]): boolean {
  try {
    const nonExportableFunctions = iterateAstNodes(parseTokens(tokens)).filter(
      (ast) => ast.type === "FUNCALL" && !functions[ast.value.toUpperCase()]?.isExported
    );
    return nonExportableFunctions.length === 0;
  } catch (error) {
    return false;
  }
}

export function getFunctionsFromTokens(tokens: Token[], functionNames: string[]) {
  // Parsing is an expensive operation, so we first check if the
  // formula contains one of the function names
  if (!tokens.some((t) => t.type === "SYMBOL" && functionNames.includes(t.value.toUpperCase()))) {
    return [];
  }
  let ast: AST | undefined;
  try {
    ast = parseTokens(tokens);
  } catch {
    return [];
  }
  return getFunctionsFromAST(ast, functionNames);
}

function getFunctionsFromAST(ast: AST, functionNames: string[]) {
  return iterateAstNodes(ast)
    .filter((node) => node.type === "FUNCALL" && functionNames.includes(node.value.toUpperCase()))
    .map((node: ASTFuncall) => ({
      functionName: node.value.toUpperCase(),
      args: node.args,
    }));
}
