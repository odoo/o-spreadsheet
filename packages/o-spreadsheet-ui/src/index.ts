import * as Engine from "@odoo/o-spreadsheet-engine";

export function salut() {
  console.log("Salut !");
}

export function createSpreadsheetModel() {
  return Engine.SpreadsheetModel();
}