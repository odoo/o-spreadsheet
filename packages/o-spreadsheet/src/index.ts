import * as UI from "@odoo/o-spreadsheet-ui";


export { UI };