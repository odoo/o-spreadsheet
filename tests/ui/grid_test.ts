import { GridModel } from "../../src/model";
import { makeTestFixture, triggerMouseEvent, GridParent, nextTick } from "../helpers";
jest.mock("../../src/ui/content_editable_helper");

let fixture: HTMLElement;
let model: GridModel;
let parent: GridParent;

beforeEach(async () => {
  fixture = makeTestFixture();
  model = new GridModel();
  model.state.viewport = { left: 0, top: 0, right: 9, bottom: 9 };
  parent = new GridParent(model);
  await parent.mount(fixture);
});

afterEach(() => {
  fixture.remove();
});

describe("Grid component", () => {
  test("can render a sheet with a merge", async () => {
    // select B2:B3 and merge
    model.selectCell(1, 1);
    model.updateSelection(1, 2);
    model.merge();
    expect(fixture.querySelector("canvas")).toBeDefined();
  });

  test("can click on a cell to select it", async () => {
    model.setValue("B2", "b2");
    model.setValue("B3", "b3");
    triggerMouseEvent("canvas", "mousedown", 300, 200);
    expect(model.state.activeXc).toBe("C8");
  });

  test("can click on resizer, then move selection with keyboard", async () => {
    model.setValue("B2", "b2");
    model.setValue("B3", "b3");
    triggerMouseEvent(".o-overlay", "click", 300, 20);
    document.activeElement!.dispatchEvent(
      new KeyboardEvent("keydown", { key: "ArrowDown", bubbles: true })
    );
    expect(model.state.activeXc).toBe("A2");
  });

  test("can shift-click on a cell to update selection", async () => {
    model.setValue("B2", "b2");
    model.setValue("B3", "b3");
    triggerMouseEvent("canvas", "mousedown", 300, 200, { shiftKey: true });
    expect(model.state.selection.zones[0]).toEqual({
      top: 0,
      left: 0,
      bottom: 7,
      right: 2
    });
  });

  describe("keybindings", () => {
    test("pressing ENTER put current cell in edit mode", async () => {
      // note: this behavious is not like excel. Maybe someone will want to
      // change this
      parent.grid.el.dispatchEvent(new KeyboardEvent("keydown", { key: "Enter" }));
      expect(model.state.activeXc).toBe("A1");
      expect(model.state.isEditing).toBe(true);
    });

    test("pressing ENTER in edit mode stop editing and move one cell down", async () => {
      model.startEditing("a");
      await nextTick();
      fixture
        .querySelector("div.o-composer")!
        .dispatchEvent(new KeyboardEvent("keydown", { key: "Enter" }));
      expect(model.state.activeXc).toBe("A2");
      expect(model.state.isEditing).toBe(false);
      expect(model.state.cells["A1"].content).toBe("a");
    });

    test("pressing shift+ENTER in edit mode stop editing and move one cell up", async () => {
      model.selectCell(0, 1);
      expect(model.state.activeXc).toBe("A2");
      model.startEditing("a");
      await nextTick();
      fixture
        .querySelector("div.o-composer")!
        .dispatchEvent(new KeyboardEvent("keydown", { key: "Enter", shiftKey: true }));
      expect(model.state.activeXc).toBe("A1");
      expect(model.state.isEditing).toBe(false);
      expect(model.state.cells["A2"].content).toBe("a");
    });

    test("pressing shift+ENTER in edit mode in top row stop editing and stay on same cell", async () => {
      model.startEditing("a");
      await nextTick();
      fixture
        .querySelector("div.o-composer")!
        .dispatchEvent(new KeyboardEvent("keydown", { key: "Enter", shiftKey: true }));
      expect(model.state.activeXc).toBe("A1");
      expect(model.state.isEditing).toBe(false);
      expect(model.state.cells["A1"].content).toBe("a");
    });

    test("pressing TAB move to next cell", async () => {
      parent.grid.el.dispatchEvent(new KeyboardEvent("keydown", { key: "Tab" }));
      expect(model.state.activeXc).toBe("B1");
    });

    test("pressing shift+TAB move to previous cell", async () => {
      model.selectCell(1, 0);
      expect(model.state.activeXc).toBe("B1");
      parent.grid.el.dispatchEvent(new KeyboardEvent("keydown", { key: "Tab", shiftKey: true }));
      expect(model.state.activeXc).toBe("A1");
    });

    test("can undo/redo with keyboard", async () => {
      model.setStyle({ fillColor: "red" });
      expect(model.state.cells.A1.style).toBeDefined();
      document.activeElement!.dispatchEvent(
        new KeyboardEvent("keydown", { key: "z", ctrlKey: true, bubbles: true })
      );
      expect(model.state.cells.A1).not.toBeDefined();
      await nextTick();
      document.activeElement!.dispatchEvent(
        new KeyboardEvent("keydown", { key: "y", ctrlKey: true, bubbles: true })
      );
      expect(model.state.cells.A1.style).toBeDefined();
    });

    test("can undo/redo with keyboard (uppercase version)", async () => {
      model.setStyle({ fillColor: "red" });
      expect(model.state.cells.A1.style).toBeDefined();
      document.activeElement!.dispatchEvent(
        new KeyboardEvent("keydown", { key: "Z", ctrlKey: true, bubbles: true })
      );
      expect(model.state.cells.A1).not.toBeDefined();
      await nextTick();
      document.activeElement!.dispatchEvent(
        new KeyboardEvent("keydown", { key: "Y", ctrlKey: true, bubbles: true })
      );
      expect(model.state.cells.A1.style).toBeDefined();
    });

    test("can select all the sheet with CTRL+A", async () => {
      document.activeElement!.dispatchEvent(
        new KeyboardEvent("keydown", { key: "A", ctrlKey: true, bubbles: true })
      );
      expect(model.state.activeXc).toBe("A1");
      expect(model.state.selection.zones[0]).toEqual({ left: 0, top: 0, right: 25, bottom: 99 });
    });

    test("can save the sheet with CTRL+S", async () => {
      let saveContentCalled = false;
      parent.el!.addEventListener("save-content", () => {
        saveContentCalled = true;
      });
      document.activeElement!.dispatchEvent(
        new KeyboardEvent("keydown", { key: "S", ctrlKey: true, bubbles: true })
      );
      expect(saveContentCalled).toBe(true);
    });
  });

  describe("paint format tool with grid selection", () => {
    test("can paste format with mouse", async () => {
      model.setValue("B2", "b2");
      model.selectCell(1, 1);
      model.setStyle({ bold: true });
      model.copy({ onlyFormat: true });
      triggerMouseEvent("canvas", "mousedown", 300, 200);
      expect(model.state.cells.C8).not.toBeDefined();
      triggerMouseEvent("body", "mouseup", 300, 200);
      expect(model.state.cells.C8.style).toBe(2);
    });

    test("can paste format with key", async () => {
      model.setValue("B2", "b2");
      model.selectCell(1, 1);
      model.setStyle({ bold: true });
      model.copy({ onlyFormat: true });
      expect(model.state.cells.C2).not.toBeDefined();
      document.activeElement!.dispatchEvent(
        new KeyboardEvent("keydown", { key: "ArrowRight", bubbles: true })
      );
      expect(model.state.cells.C2.style).toBe(2);
    });
  });
});
