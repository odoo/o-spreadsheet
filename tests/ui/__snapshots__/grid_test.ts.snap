// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid component simple rendering snapshot 1`] = `
<div
  class="o-grid"
>
  <canvas
    height="150"
    style="width:-15px;height:-15px;"
    tabindex="-1"
    width="300"
  />
  <div
    class="o-overlay"
  >
    <div
      class="o-col-resizer"
    />
    <div
      class="o-row-resizer"
    />
    <div
      class="all"
    />
  </div>
  <div
    class="o-scrollbar vertical"
  >
    <div
      style="width:1px;height:2328px"
    />
  </div>
  <div
    class="o-scrollbar horizontal"
  >
    <div
      style="height:1px;width:2592px"
    />
  </div>
</div>
`;
